R/code_feedback.R

```diff
@@ -218,6 +234,52 @@ with_maybe_code_feedback <- function(val, expr) {
   )
 }
 
+which_closest_solution_code <- function(user_code, solution_code_all) {
+  # If there's no solution code or only one solution,
+  # we don't need to find the closest match
+  if (length(solution_code_all) < 2) {
+    return(length(solution_code_all))
+  }
+
+  # Convert from list to character vector
+  solution_code_all <- unlist(solution_code_all)
+
+  standardise_code_text <- function(code) {
+    code %>%
+      unpipe_all_str() %>%
+      rlang::parse_exprs() %>%
+      call_standardise_formals_recursive() %>%
+      purrr::map_chr(rlang::expr_text) %>%
+      paste(collapse = "\n")
+  }
+
+  user_code <- standardise_code_text(user_code)
+
+  solution_code_all <- solution_code_all %>%
+    purrr::map_chr(standardise_code_text)
+
+  # Find the index of the solution code that the user code is closest to
+  # which.min.last() uses the last index if there is a tie
+  closest_solution <- which.min.last(utils::adist(user_code, solution_code_all))
+
+  # If closest_solution is invalid,
+  # fallback to the last element of solution_code_all
+  if (
+    length(closest_solution) != 1 ||
+      !closest_solution %in% seq_along(solution_code_all)
+  ) {
+    closest_solution <- length(solution_code_all)
+  }
+
+  closest_solution
+}
+
+which.min.last <- function(x) {
+  x <- rev(x)
+  index <- which.min(x)
+  rev(seq_along(x))[index]
```

Comment:
Is there an advantage to using `which.min()`? This is more verbose but less mind-bendy with the reversing:

```r
idx_min <- which(x == min(x))

if (length(idx_min) < 2) {
  return(idx_min)
}

if (length(x) %in% idx_min) {
  # return last solution when it's in the tie
  return(length(x))
}

idx_min[1]
```

Also the version above returns the _final_ solution if it's in the tie, not the last solution _in the tie_. Basically: we take the final solution whenever it's in the possibilities because it's the default solution. But if it's not the final solution we'd use the first one on the list.

Reply:
That's different enough from `which.min.last()` that maybe we'd want to move it back into `which_closest_solution_code()`.

Reply:
This seems like a good solution. The advantage of `which.min()` is that it's very fast, but exercise evaluation is a pretty slow process anyway, so I don't think shaving off a few milliseconds is worth less readable code.

Reply:
I implemented this in [5c11542](https://github.com/rstudio/gradethis/pull/289/commits/5c1154253e871364e291edbf1362519c18fc8f32). The result is now:
1. Tie for all solutions – use last
2. Tie among some solutions, including the last solution – use last
3. Tie among some solutions, not including the last solution – use earliest solution included in tie
4. Clear winner (by any amount) – use winner